numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -1134,6 +1134,9 @@ static void
 npyiter_dealloc(NewNpyArrayIterObject *self)
 {
     if (self->iter) {
+        /* Store error, so that WriteUnraisable cannot clear an existing one */
+        PyObject *exc, *val, *tb;
+        PyErr_Fetch(&exc, &val, &tb);
```

Comment:
Surprisingly, there was a test that actually tripped when `dealloc` accidentally cleared an existing error.  Apparently, that is important, at least if we fail during `init`.